osu.Game/Screens/SelectV2/SongSelect.cs

```diff
@@ -569,6 +571,24 @@ private void queueBeatmapSelection(GroupedBeatmap groupedBeatmap)
             debounceQueueSelection(groupedBeatmap.Beatmap);
         }
 
+        private bool checkAndApplyRuleset()
+        {
+            if (Beatmap.Value is DummyWorkingBeatmap || Beatmap.Value.BeatmapInfo.AllowGameplayWithRuleset(Ruleset.Value, showConvertedBeatmaps.Value)) return false;
+
+            // try to find a beatmap whose ruleset is current.
+            var beatmapWithinCurrentRuleset = Beatmap.Value.BeatmapSetInfo.Beatmaps.FirstOrDefault(b => b.Ruleset.Equals(Ruleset.Value));
+
+            if (beatmapWithinCurrentRuleset != null)
+            {
+                Beatmap.Value = beatmaps.GetWorkingBeatmap(beatmapWithinCurrentRuleset);
```

Comment:
why does this unconditionally touch the global beatmap again? inside a global beatmap change callback? that's recursive. why doesn't it only perform the change when necessary?